### DISCUSSION TITLE
Modal Dialogs?

### DISCUSSION THREAD

**Discussion by npwoods** [slint-ui/slint#6028] [Q&A]
Is there a canonical technique to create modal dialogs in Slint?

I can't speak for other platforms, but on Windows this requires passing specific parameters to the API that creates a window (`CreateWindowEx`), and I suspect that this might be quite gross if I tried to do it on my own.

**Comment by Horbin-Magician** [comment]
You can use PopupWindow, ref: https://releases.slint.dev/1.7.2/docs/slint/src/language/builtins/elements.html#popupwindow.
Besides, Slint now supports multiple Windows, so you can new a window as the modal dialog instead of using CreateWindowEx.

**Comment by npwoods** [comment]
`PopupWindow` is inadequate; it appears to be the equivalent of a HTML popup implemented with a span tag.

I want to have a modal popup that is equivalent to what you get in WIndows `notepad.exe` if you try to exit without saving changes.  Unless I'm mistaken, this cannot be achieved in Slint without overriding the `winit`'s creation of the Window because in Win32, you have to specify the modal parent as the `hwndOwner` parameter to `CreateWindowEx`.

**Comment by npwoods** [comment]
[modal-hack-demo.zip](https://github.com/user-attachments/files/16922687/modal-hack-demo.zip)

I was able to hack something together that will create a real Win32 modal dialog with Slint that properly blocks the underlying Window.  I'm hoping that there will be a better way to do this because:
* This solution is Windows specific
* The modal dialog is not centered on the parent properly
* When the modal dialog is closed, focus to the underlying window seems to lost
* This is a disgusting hack

**Reply by Enyium** [reply]
There's also [`SetParent()`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setparent) that allows you to set the owner, not just `CreateWindowExW()`. [More on owned windows.](https://learn.microsoft.com/en-us/windows/win32/winmsg/window-features#owned-windows) And this article from Raymond Chen is also relevant: [The correct order for disabling and enabling windows](https://devblogs.microsoft.com/oldnewthing/20040227-00/?p=40463).

**Comment by npwoods** [comment]
Yes, that is extremely relevant - my hacky solution creates an object whose sole purpose is to reenable the window (`WindowReenabler`) when dropped, and attaches it to the window as a bogus rendering notifier with `set_rendering_notifier()`.  Honestly, I'm not sure if this gets the ordering right nor not.  And I really don't want to be doing platform specific stuff just to show a modal dialog.
